File: day1/day1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <unordered_map>
#include <set>
#include <cctype>
#include <algorithm>

class Day1 {


public:
    std::vector<std::string> lines;
    std::unordered_map<std::string, int> wordsToInt{{"one",   1},
                                                    {"two",   2},
                                                    {"three", 3},
                                                    {"four",  4},
                                                    {"five",  5},
                                                    {"six",   6},
                                                    {"seven", 7},
                                                    {"eight", 8},
                                                    {"nine",  9}};
    std::vector<std::string> words = {"one",
                                      "two",
                                      "three",
                                      "four",
                                      "five",
                                      "six",
                                      "seven",
                                      "eight",
                                      "nine"};

    Day1() {
        lines = getLines();
    }

    std::vector<std::string> getLines() {
        std::ifstream myfile("input.txt");
        std::string line;
        std::vector<std::string> input;
        while (std::getline(myfile, line))
        {
            input.push_back(line);
        }
        return input;
    }

    int part1() {
        std::vector<std::vector<int>> processedLines = preprocessStringsPart1();
        return sumOuterNumbers(processedLines);
    }

    int part2() {
        std::vector<std::vector<int>> processedLines = preprocessStringsPart2();
        return sumOuterNumbers(processedLines);
    }

    std::vector<std::vector<int>> preprocessStringsPart1() {
        std::vector<std::vector<int>> processedLines{};
        std::vector<int> processedLine{};
        for (const auto& line : lines) {
            processedLine = {};
            for (const auto& c : line) {
                if (std::isdigit(c)) {
                    processedLine.push_back(static_cast<int>(c - '0'));
                }
            }
            processedLines.push_back(processedLine);
        }
        return processedLines;
    }

    std::vector<std::vector<int>> preprocessStringsPart2() {
        std::vector<std::vector<int>> processedLines{};
        int l = 0;
        for (const auto& line : lines) {
            std::vector<int> processedLine{};
            while (l < line.size()) {
                if (std::find(words.begin(), words.end(), line[l]) != words.end())
                    processedLine.push_back(static_cast<int>(line[l] - '0'));
                else {
                    int answer = isStringRangeValid(line, l);
                    if (answer != 0) 
                        processedLine.push_back(answer);
                }
                l++;
            }
            return processedLines;
        }
    }

    int isStringRangeValid(std::string line, int l) {
        int r = l + 3;
        while (r - l <= 5 and r < line.size() + 1) {
            if (wordsToInt.find( line.substr(l, r - l + 1) ) != wordsToInt.end())
                return wordsToInt.at(line.substr(l, r - l + 1));
            r++;
        }
        return 0;
    }

    int sumOuterNumbers(std::vector<std::vector<int>> processedLines) {
        int sum{};
        for (const auto& line : processedLines) {
            sum += 10 * line.front();
            sum += line.back();
        }
        return sum;
    }

};

int main(int argc, char* argv[]) {

    Day1 day1{};

    int result = day1.part1();
    std::cout << "Task 1: " << result << std::endl;
    
    return 1;
}

File: day1/day1.py
import string
import sys


class Day1:

    char_string_map = {"one"   : 1,
                       "two"   : 2,
                       "three" : 3,
                       "four"  : 4,
                       "five"  : 5,
                       "six"   : 6,
                       "seven" : 7,
                       "eight" : 8,
                       "nine"  : 9,
                       }
    min_char_string_length = min([len(key) for key in char_string_map])
    max_char_string_length = max([len(key) for key in char_string_map])

    inverse_char_string_map = {k[::-1] : v for k, v in char_string_map.items()}

    def __init__(self):
        self.input = Day1.get_input()

    def part_1(self) -> int:
        lines = Day1.preprocess_strings(self.input, Day1.string_preprocessing_part_1)
        return Day1.handle_boundary_integers(lines)
    
    def part_2(self) -> int:
        lines = Day1.preprocess_strings(self.input, Day1.string_preprocessing_part_2)
        return Day1.handle_boundary_integers(lines)

    @staticmethod
    def handle_boundary_integers(lines: list[list[int]]) -> int:
        assert len(lines) >= 1 and all(all((type(item) == int) for item in line) for line in lines)
        value = 0
        for line in lines:
            value += 10 * line[0] + line[-1]
        return value
    
    @staticmethod
    def preprocess_strings(lines: list[str], method) -> list[list[int]]:
        new_lines = []
        for line in lines:
            new_lines.append(method(line))
        return new_lines
    
    @staticmethod
    def string_preprocessing_part_1(line: str) -> list[list[int]]:
        new_line = []
        for c in line:
            if c.isdigit():
                new_line.append(int(c))
        return new_line
    
    @staticmethod
    def string_preprocessing_part_2(line: str) -> list[int]:
        new_line = []
        l = 0
        while l < len(line):
            if line[l] in string.digits:
                new_line.append(int(line[l]))
            else:
                answer = Day1.isStringRangeValid(line, l)
                if answer:
                    new_line.append(answer)
            l += 1
        return new_line
    
    @staticmethod
    def isStringRangeValid(line, l):
        r = l + Day1.min_char_string_length
        while r - l <= Day1.max_char_string_length and r < len(line) + 1:
            if line[l:r] in Day1.char_string_map:
                return Day1.char_string_map[line[l:r]]
            r += 1
        return 0

    @staticmethod
    def get_input() -> list[str]:
        with open("input.txt", 'r') as f:
            lines = f.read().splitlines()
        return lines

if __name__ == "__main__":

    day1 = Day1()

    print("Python Solution:")
    print(f"\tPart 1: {day1.part_1()}")
    print(f"\tPart 2: {day1.part_2()}")
    